### PR TITLE
Add synthetic source support for geo_shape via fallback implementation

### PR DESCRIPTION
**Pull request by lkts**
This PR enables `geo_shape` mapper to use fallback synthetic source infrastructure and as such adds synthetic source support for this field type. 

Note that `geo_shape` fields are supported by ESQL but only using source to load data. ESQL will not work for `geo_shape` fields with synthetic source, tracked by #108883.

### CODE REVIEW

File: x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapperTests.java [hunk 1]
```diff
@@ -422,7 +434,161 @@ protected Object generateRandomInputValue(MappedFieldType ft) {
 
     @Override
     protected SyntheticSourceSupport syntheticSourceSupport(boolean ignoreMalformed) {
-        throw new AssumptionViolatedException("not supported");
+        // Almost like GeoShapeType but no circles
```
Comment (kkrik-es) [inline x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapperTests.java:437]: Why do we skip circles? Also, shall we use GeoShapeType and skip the unsupported types in the switch?

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/spatial/140_synthetic_source.yml [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+            string: "string"
+            array: [{ "a": 1 }, { "b": 2 }]
+            object: { "foo": "bar" }
+
+  - do:
+      index:
+        index: test
+        id: "3"
+        body:
+          shape: ["POINT (-77.03653 38.897676)", "potato", "POINT (-71.34 41.12)"]
```
Comment (kkrik-es) [inline x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/spatial/140_synthetic_source.yml:99]: Add a test were tha point is partially valid, e.g. (-71.34, 1000) ?

**Comment by kkrik-es** [comment]
Looks good overall, adding @craigtaverner to double-check the tests.

**Review by mark-vieira — APPROVED** [review]

File: x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapperTests.java (hunk 1 above)
Comment (lkts) [reply x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapperTests.java:437]: This field does not support circles https://www.elastic.co/guide/en/elasticsearch/reference/current/geo-shape.html#input-structure.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapper.java [hunk 3]
```diff
@@ -482,6 +482,11 @@ protected void checkIncomingMergeType(FieldMapper mergeWith) {
         super.checkIncomingMergeType(mergeWith);
     }
 
+    @Override
+    protected SyntheticSourceMode syntheticSourceMode() {
```
Comment (martijnvg) [inline x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapper.java:486]: Note that it should be relatively easy to natively support synthetic source for `GeoShapeWithDocValuesFieldMapper`, because this field mapper can store shapes in an accurate and more efficient format as well. We should then default the `store` attribute to `true`, like we did for other field types as well. We can do this in a follow up.

File: modules/parent-join/build.gradle [hunk 4]
```diff
@@ -16,7 +16,7 @@ esplugin {
 
 restResources {
   restApi {
-    include '_common', 'bulk', 'cluster', 'nodes', 'indices', 'index', 'search'
+    include '_common', 'bulk', 'cluster', 'get', 'nodes', 'indices', 'index', 'search'
```
Comment (martijnvg) [inline modules/parent-join/build.gradle:19]: Why is this change needed? The new yaml tests does do get api requests, but that is in a different module then this change is.

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/spatial/140_synthetic_source.yml (hunk 2 above)
Comment (martijnvg) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/spatial/140_synthetic_source.yml:99]: 👍 (basically a yml test with ignore malformed)

**Review by martijnvg — APPROVED** [review]
Left two comments. LGTM otherwise.

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapper.java (hunk 3 above)
Comment (lkts) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapper.java:486]: One benefit of current implementation is that it covers `ignore_malformed` automatically. I'll create an issue and we can prioritize it based on usage.

Comment (lkts) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapper.java:486]: #108980

**Comment by lkts** [comment]
buildkite test this please

File: x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapperTests.java (hunk 1 above)
Comment (kkrik-es) [reply x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapperTests.java:437]: I'd still use `GeoShapeType` and map the behavior for circle to a valid one, e.g. point. That way, this won't get out of sync if another type is added to `GeoShapeType`. Mostly nit.

**Review by kkrik-es — APPROVED** [review]

File: x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapper.java (hunk 3 above)
Comment (martijnvg) [reply x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/GeoShapeWithDocValuesFieldMapper.java:486]: 👍
